### PR TITLE
added debug argument to ng test task.

### LINKED ISSUE #1798: Source Code showed in Browser Developer Tools is showing instrumented source code.
> Please provide us with the following information:
> ---------------------------------------------------------------

1. OS?  Windows 7, 8 or 10. Linux (which distribution). Mac OSX (Yosemite? El Capitan?)
Ubuntu 16

2. Versions. Please run `ng --version`. If there's nothing outputted, please run
in a Terminal: `node --version` and paste the result here:

angular-cli: local (v1.0.0-beta.11-webpack.2, branch: master)
node: 6.3.0
os: linux x64

3. Repro steps. Was this an app that wasn't created using the CLI? What change did you
   do on your code? etc.

When trying to debug my tests the source code mapped is showing illegible because of the instrumented code needed for the coverage report feature.

As seen in the screenshot bellow:

I think we could have an argument `debug` which we would pass to the ng test task which will not run the coverage config parts and, this way, debug will be possible.

I already have this implemented here and will submit a merge request for that.

4. The log given by the failure. Normally this include a stack trace and some
   more information.

5. Mention any other details that might be useful.

> ---------------------------------------------------------------
> Thanks! We'll be in touch soon.

### PR DESCRIPTION
**Pull request by abner**
added debug argument to ng test task. This argument when passed disable instrumentation and coverage, allowing to debug the source code when running tests. Fixes #1798

---
This change is [Reviewable](https://reviewable.io/reviews/angular/angular-cli/1799)

### CODE REVIEW

**Comment by abner** [comment]
@filipesilva I added this pull request to solve an issue related to not being able to see application code in debug karma session  because of the instrumentation made to report the coverage. Basically i added a debug argument to ng test (`ng test debug`) which disable coverage and instrumentation parts and this way i could see my application code in chrome developer tool debug.
Could you review this and see if it would be merged.
Thanks in advance!

**Comment by filipesilva** [comment]
Before I review, can I ask you to test in the latest master? We had a problem with sourcemaps that is fixed I believe, and would cause you to see instrumented code.

@TheLarkInn can you have a look at https://github.com/angular/angular-cli/issues/1798 and tell me if this is the sourcemaps issue?

**Comment by TheLarkInn** [comment]
No this is specifically related to instrumented code in testing mode.

**Comment by abner** [comment]
I reviewed and it still occurs. As @TheLarkInn said it is related to instrumented code.

**Comment by abner** [comment]
hi @TheLarkInn and @filipesilva !
I still need this feature in angular-cli. I updated the pull request with master.
Could you guys check if it would be merged?
Thanks in advance.

**Comment by filipesilva** [comment]
@abner I'm sorry it's taken so long to review this PR. I've bumped the original issue as something that needs to be finished for RC1 and we should get to it soon.

**Comment by abner** [comment]
great @filipesilva . I just updated the branch.

**Comment by filipesilva** [comment]
CI is showing a lint error:
```
/home/travis/build/angular/angular-cli/packages/angular-cli/models/webpack-build-test.js
  9:7  error  'debug' is already defined  no-redeclare
```
It's minor but I cannot merge it without a green CI.

File: packages/angular-cli/commands/test.ts [hunk 1]
```diff
@@ -5,6 +5,7 @@ import {CliConfig} from '../models/config';
 const NgCliTestCommand = TestCommand.extend({
   availableOptions: [
     { name: 'watch', type: Boolean, default: true, aliases: ['w'] },
+    { name: 'debug', type: Boolean, default: false, aliases: ['d'] },
```
Comment (filipesilva) [inline packages/angular-cli/commands/test.ts]: How is this option being passed into the karma config? I don't think it's being passed in yet.

My suggestion is this: in `karma.conf.js` add in the debug property to the `angularCli` options:
```
    angularCli: {
      config: './angular-cli.json',
      environment: 'dev',
      debug: false
    },
```

Then you'll need to also change `packages/angular-cli/tasks/test.ts` to set this to `true` in the `options` variable that's being passed into `karma.Server`.

Make sure it works though, this is the kind of thing I don't know we can really do CI tests without doing really weird stuff.

File: packages/angular-cli/models/webpack-build-test.js [hunk 2]
```diff
@@ -66,7 +67,7 @@ const getWebpackTestConfig = function (projectRoot, environment, appConfig) {
         { test: /\.(jpg|png)$/, loader: 'url-loader?limit=128000' },
         { test: /\.html$/, loader: 'raw-loader', exclude: [path.resolve(appRoot, appConfig.index)] }
       ],
-      postLoaders: [
+      postLoaders: debug ? [] : [
```
Comment (filipesilva) [inline packages/angular-cli/models/webpack-build-test.js]: In the latest webpack update, the loader configs changed a bit and is now a single `rules` array. 

You're probably going to have to add the `sourcemap-istanbul-instrumenter-loader` conditionally, but by itself, at the end.

File: packages/angular-cli/models/webpack-build-test.js [hunk 3]
```diff
@@ -3,9 +3,10 @@
 const path = require('path');
 const webpack = require('webpack');
 
-const getWebpackTestConfig = function (projectRoot, environment, appConfig) {
+const getWebpackTestConfig = function (projectRoot, environment, appConfig, debug) {
 
   const appRoot = path.resolve(projectRoot, appConfig.root);
+  var debug = typeof debug !== 'undefined' ?  debug : false;
```
Comment (filipesilva) [inline packages/angular-cli/models/webpack-build-test.js]: Linting says this variable is being shadowed.

File: packages/angular-cli/models/webpack-build-test.js (hunk 2 above)
Comment (abner) [inline packages/angular-cli/models/webpack-build-test.js]: ok @filipesilva . i will update this branch to the latest changes in master and fix the linting issue you mentioned.

File: packages/angular-cli/commands/test.ts (hunk 1 above)
Comment (abner) [inline packages/angular-cli/commands/test.ts]: i actually implemented debug false as default so it doesn't need to be passed. i added an argument to the test command --debug (-d). I think we would add a npm script npm run test:debug which would  call ng test --debug

**Comment by filipesilva** [comment]
@abner we were talking about this PR between the team, and thinking of a somewhat alternative solution. 

How about having a `--code-coverage` (`-cc`) flag that adds the instrumentation, instead of having a `--debug` flag to disable it? I don't think users care about code coverage during their normal watch cycle, and it should make things faster as well.


What do you think?

**Pull request closed without merging**